### PR TITLE
[RFC][distributed] Add parallel_map with collective operation interception…

### PR DESCRIPTION
**Pull request by leveretconey**
Relates to https://github.com/pytorch/pytorch/issues/180550

## Motivation

Under free-threaded Python (nogil), user code that runs independent iterations in parallel threads can easily hit collective ordering issues — when multiple threads call `dist.all_reduce` (or other collectives) concurrently, cross-rank pairing becomes non-deterministic, leading to silent data corruption or hangs.

Users could in principle restructure their code to move collective calls out of the parallel region, but this becomes impractical when collectives are buried deep inside third-party libraries or framework internals that the user cannot easily access or modify. This PR provides a framework-level solution that minimizes user-side changes: just replace the for-loop with `parallel_map`, and the user function itself does not need to change at all.

The primary use case is parallelizing serial for-loops in training (e.g., per-decoder-layer loss computation) where each iteration contains both CPU-side Python logic and GPU/distributed ops. See `test/distributed/test_parallel_mapper.py` for detailed usage examples.

## Summary

Implements `parallel_map` / `parallel_starmap` / `parallel_multi_apply` APIs that enable multi-threaded parallel execution under free-threaded Python, with automatic serialization of `torch.distributed` collective operations.

```python
# Before
results = [work(x) for x in inputs]

# After
from torch.distributed.parallel import parallel_map
results = parallel_map(work, inputs)
```

## Design

The key idea is intercepting collective operations at the Python API level in `distributed_c10d.py`. When a worker thread calls `dist.all_reduce` (or any collective), the call is redirected to the main thread and executed in strict worker-id order, ensuring cross-rank correctness.

```
Workers (parallel):  [compute] → [collective intercepted] → [compute]
Main thread:                      serializes in order 0, 1, 2, ...
```

The interceptor check is a lightweight thread-local attribute lookup (`getattr(thread_local, ..., None)`), so there is effectively zero overhead when `parallel_map` is not in use.

Components added:
- `torch/distributed/_collective_interceptor.py` — thread-local interceptor primitives with re-entrancy guard
- `torch/distributed/parallel/_parallel_mapper.py` — `_SyncCoordinator`, `_WorkerThread`, `_ParallelPool`
- `torch/distributed/parallel/__init__.py` — public API (`parallel_map`, `parallel_starmap`, `parallel_multi_apply`, `sync_wrap`)
- `test/distributed/test_parallel_mapper.py` — test suite
- `torch/distributed/distributed_c10d.py` — interceptor checks at ~18 collective entry points

## This is a preliminary implementation

We're sharing this early implementation to **get feedback on the design approach**:

- Is intercepting collectives at the Python API level the right place, or should this be done at a lower layer (ProcessGroup / C++)?
- Does the main-thread serialization model make sense?
- What should the public API look like?

If the overall direction is well-received, we'll follow up with API refinement, additional tests, and documentation based on community feedback.

### CODE REVIEW

**Comment by malfet** [comment]
Please continue discussion on the issue, and if one of the folks on distributed team will ask you to raise the PR, mention this one
